### ISSUE TITLE
Eubos v2.24 blunder : cannot reproduce - Transposition Table corruption on shutdown.

### ISSUE THREAD

**Issue by tissatussa** [cjbolt/EubosChess#10]
lately i was testing some chess position with several engines in CuteChess and let them play a game with a rather long time control (tc 780+10) to see how they would approach this position : Eubos v2.24 failed in a strange way ..

it was a Dutch Defence : 1.d4 f5 2.Qd3!?
`rnbqkbnr/ppppp1pp/8/5p2/3P4/3Q4/PPP1PPPP/RNB1KBNR b KQkq - 1 2`

at one point in the game Eubos didn't save its piece, lost that piece and thus lost the game :

![eubos_v2 24_wrong_move](https://github.com/cjbolt/EubosChess/assets/1109281/00035feb-35cd-402e-958e-9c95332ebf1e)

here Eubos v2.24 (White) played 13.Nd3?? after 15 seconds, reaching search depth 15 .. when i try to reproduce this outcome in terminal, i don't succeed :

```
$ java -jar /home/roelof/Apps/ChessEngines/0_JARs/Eubos/v2.24/Eubos-v2.24-asset.jar
uci
id name Eubos 2.24
id author Chris Bolt
option name Hash type spin default 256 min 32 max 4000
option name Threads type spin default 2 min 1 max 4
option name Move Overhead type spin default 10 min 0 max 5000
option name Lazy Threshold type spin default 275 min 0 max 1000
uciok
ucinewgame
position fen r1bqkbr1/1p5p/4p3/1p1pNpp1/2pPn3/5PPB/PPP1P2P/R2Q1RK1 w q - 0 14
go infinite
info depth 1 seldepth 5 score cp -33 pv f3e4 d5e4 e5c4 hashfull 0 nps 347 time 46 nodes 46
info depth 2 seldepth 5 score cp -33 pv f3e4 d5e4 e5c4 hashfull 0 nps 3274 time 33 nodes 114
info depth 3 seldepth 10 score cp -39 pv f3e4 g5g4 h3g2 d5e4 hashfull 0 nps 19980 time 112 nodes 1456
info depth 4 seldepth 14 score cp -40 pv f3e4 d5e4 d1d2 d8a5 d2a5 a8a5 hashfull 0 nps 29074 time 173 nodes 4031
info depth 5 seldepth 13 score cp -42 pv f3e4 d5e4 h3g2 f8d6 c2c3 hashfull 0 nps 60495 time 206 nodes 11180
info depth 6 seldepth 18 score cp -34 pv f3e4 d5e4 e2e3 g5g4 h3g2 f8b4 e5g4 hashfull 0 nps 150013 time 350 nodes 39361
info depth 7 seldepth 15 score cp -31 pv f3e4 d5e4 e2e3 g5g4 h3g2 f8d6 a2a3 d6e5 d4e5 hashfull 0 nps 214045 time 443 nodes 61586
info depth 8 seldepth 22 score cp -31 pv f3e4 d5e4 e2e3 g5g4 h3g2 hashfull 0 nps 240895 time 648 nodes 123262
info depth 9 seldepth 20 score cp -35 pv f3e4 d5e4 e2e3 g5g4 h3g2 f8d6 a2a3 d8c7 d1d2 d6e5 d4e5 c7e5 hashfull 1 nps 256606 time 910 nodes 199517
info depth 10 seldepth 22 score cp -56 pv f3e4 d5e4 e2e3 g5g4 h3g2 f8d6 a2a3 d6e5 hashfull 2 nps 297999 time 1355 nodes 342402
info depth 11 seldepth 23 score cp -52 pv f3e4 d5e4 e2e3 g5g4 h3g2 f8g7 a2a3 d8d5 c2c3 g7e5 d4e5 d5d1 hashfull 5 nps 344499 time 2385 nodes 752096
info depth 12 seldepth 29 score cp -79 pv f3e4 d5e4 a2a3 f8d6 e2e3 g5g4 h3g2 d6e5 d4e5 d8d1 a1d1 c8d7 d1d7 hashfull 17 nps 542047 time 5232 nodes 2774678
info depth 13 seldepth 25 score cp -76 pv f3e4 d5e4 a2a3 e4e3 g3g4 f8g7 g4f5 g5g4 f5f6 g7f6 h3g2 g4g3 h2g3 g8g3 hashfull 33 nps 683782 time 7357 nodes 4900476
info depth 14 seldepth 29 score cp -87 pv f3e4 d5e4 a2a3 e4e3 g3g4 f8g7 c2c3 g7e5 d4e5 d8d1 a1d1 hashfull 52 nps 825189 time 9782 nodes 7739745
info depth 15 seldepth 30 score cp -87 pv f3e4 d5e4 a2a3 d8d5 d1d2 f8d6 d2c3 b7b6 h3g2 a8a7 a1d1 g8g7 e5d3 e8f7 d3e5 d6e5 d4e5 hashfull 103 nps 907314 time 15935 nodes 14030114
info depth 16 seldepth 31 score cp -93 pv f3e4 d5e4 h3g2 f8d6 e2e3 d6e5 d4e5 d8d1 f1d1 c8d7 g3g4 e8e7 g4f5 e6f5 d1d6 b5b4 d6d7 hashfull 203 nps 1067545 time 24704 nodes 25992431
info depth 17 seldepth 33 score cp -107 pv f3e4 d5e4 h3g2 f8d6 e2e3 g5g4 a2a3 d8c7 c2c3 d6e5 d4e5 c7e5 hashfull 332 nps 1179899 time 34939 nodes 40687807
info depth 18 seldepth 34 score cp -94 pv f3e4 d5e4 h3g2 f8d6 e2e3 g5g4 d1e1 d8c7 e1c3 c7a5 c3a5 a8a5 a2a3 c8d7 e5d7 e8d7 d4d5 a5a8 d5e6 hashfull 657 nps 1209966 time 61902 nodes 74352415
info depth 19 seldepth 39 score cp -95 pv f3e4 d5e4 h3g2 f8d6 e2e3 g5g4 d1e1 d8c7 e1c3 c7a5 c3a5 a8a5 c2c3 d6e5 hashfull 994 nps 1263894 time 137974 nodes 173784923
info depth 20 seldepth 38 score cp -95 pv f3e4 g5g4 h3g2 d5e4 c2c3 f8d6 d1c1 d6e5 d4e5 d8c7 f1d1 c7c5 hashfull 1000 nps 1298784 time 479732 nodes 622690576
```

(i removed all info lines without a score)

this test in terminal shows Eubos v2.24 wants to capture the Black kNight by f3xe4, which indeed is the only good move, obviously - i wonder why Eubos didn't do that in the game ..

can you explain / reproduce this 'behaviour' ? 

---

Here's the full game :

```
[Event "engine -vs- engine"]
[Site "Holland"]
[Date "2023.11.01"]
[Round "?"]
[White "Eubos v2.24"]
[Black "4ku v3.1"]
[Result "0-1"]
[FEN "rnbqkbnr/ppppp1pp/8/5p2/3P4/3Q4/PPP1PPPP/RNB1KBNR b KQkq - 1 2"]
[GameDuration "00:31:13"]
[PlyCount "75"]
[SetUp "1"]
[TimeControl "780+10"]

1... e6 {-0.44/24 50s} 2. Nf3 {+0.30/14 19s} Nf6 {-0.40/23 33s} 3. Nc3 {+0.23/15 63s} d5 {+0.05/25 92s} 4. g3 {+0.51/13 25s} c5 {0.00/23 25s} 5. Bh3 {+0.48/12 24s} Nc6 {+0.16/25 56s} 6. O-O {+0.24/13 27s} Ne4 {+0.23/23 21s} 7. Bf4 {+0.19/13 28s} g5 {+0.85/26 23s} 8. Be5 {+0.31/13 36s} Rg8 {+0.88/25 25s} 9. Nb5 {0.00/13 22s} Nxe5 {+1.26/25 27s} 10. Nxe5 {-0.63/13 15s} a6 {+1.25/25 20s} 11. f3 {-0.77/14 26s} c4 {+1.28/26 20s} 12. Qd1 {-1.08/15 29s} axb5 {+1.80/25 21s} 13. Nd3 {-0.97/15 15s} cxd3 {+5.92/26 23s} 14. fxe4 {-6.82/14 15s} dxe4 {+6.10/24 17s} 15. e3 {-6.57/15 29s} g4 {+5.92/26 18s} 16. Bg2 {-6.74/15 19s} dxc2 {+5.99/25 24s} 17. Qxc2 {-6.44/14 15s} b4 {+6.05/24 24s} 18. Rfd1 {-6.67/14 41s} Bd7 {+6.04/22 15s} 19. Bf1 {-6.72/14 46s} Rc8 {+6.27/22 15s} 20. Qd2 {-6.91/14 15s} h5 {+6.15/25 30s} 21. Rac1 {-6.88/14 26s} Rxc1 {+6.95/25 15s} 22. Qxc1 {-8.05/16 23s} Bd6 {+6.94/26 21s} 23. d5 {-7.06/16 23s} e5 {+7.11/23 33s} 24. Qd2 {-7.08/14 28s} h4 {+7.26/22 15s} 25. Kg2 {-7.43/14 36s} hxg3 {+8.49/23 16s} 26. hxg3 {-8.17/15 23s} Qg5 {+8.75/23 22s} 27. Re1 {-9.09/14 41s} Qh6 {+10.83/24 14s} 28. Qc2 {-9.81/16 54s} Rh8 {+17.76/30 15s} 29. Kf2 {-14.14/17 36s} Qh2+ {+18.81/29 12s} 30. Bg2 {-14.78/17 44s} Rh3 {+20.58/30 17s} 31. Re2 {-16.78/16 15s} Rxg3 {+17.93/29 110s} 32. Kf1 {-25.29/17 18s} b6 {+15.35/24 8.9s} 33. Qc4 {-15.52/13 11s} Bc5 {+18.09/28 9.5s} 34. d6 {-22.02/15 19s} Kf8 {+25.28/28 14s} 35. Qxb4 {-23.67/15 17s} Bxb4 {+327.61/28 12s} 36. Rd2 {-29.01/5 0.042s} Bb5+ {+327.63/25 10s} 37. Ke1 {-M4/1 0.023s} Qg1+ {+327.65/26 11s} 38. Bf1 {-M2/1 0.042s} Qxf1# {+327.67/27 14s, Black mates} 0-1
```

**Comment by cjbolt** [comment]
Thanks for this defect report. I too have observed some infrequent strange blunders. I think there is a defect where, either the hash table gets corrupted, or, when a search is terminated due to clock management reasons (like Eubos has spent too much time in the search) the principal variation is incorrect. Very likely the defect could be a combination of these two problems.

My conjecture is, when you re-ran in analysis mode, a search was never interrupted, which meant the search never needed to terminate whilst it was still processing, and therefore, likely the hash table didn't get corrupted.

It is valuable that you have sent me a test position from early in a game, I will look into it.

**Comment by cjbolt** [comment]
Accidentally closed

**Comment by cjbolt** [comment]
This defect is caused by writing the move currently being searched into the hash table when a search is terminated. There is one/or more code paths that don't assign a score of zero to the Transposition stored in the trans table and which allow the update to be performed. I haven't completely narrowed it down yet, but I have collected enough evidence to be able to point the finger at this failure mechanism.

At position 

`rnb2rk1/4Bpbp/p2pp1p1/qp6/4P1n1/1BN5/PPPQNPPP/R3K2R b KQ - - 14`

Eubos blundered with Bh8??

My logging shows:

```
<record>
  <date>2023-12-07T21:14:46.919114500Z</date>
  <millis>1701983686919</millis>
  <nanos>114500</nanos>
  <sequence>528</sequence>
  <logger>eubos.main</logger>
  <level>INFO</level>
  <class>eubos.search.searchers.IterativeMoveSearcher$IterativeMoveSearchStopper</class>
  <method>shouldTerminateIfEvalAboveThreshold</method>
  <thread>41</thread>
  <message>checkPoint=1 hasBackedUpAScore=true research_asp=false currentScore=-8 refScore=-46 depth=13 refDepth=11 SearchStopped=false StopperActive=true ranFor=14832</message>
</record>
<record>
  <date>2023-12-07T21:14:46.919114500Z</date>
  <millis>1701983686919</millis>
  <nanos>114500</nanos>
  <sequence>529</sequence>
  <logger>eubos.main</logger>
  <level>INFO</level>
  <class>eubos.search.searchers.MultithreadedIterativeMoveSearcher</class>
  <method>haltAllWorkers</method>
  <thread>41</thread>
  <message>Halting all workers</message>
</record>
<record>
  <date>2023-12-07T21:14:46.919114500Z</date>
  <millis>1701983686919</millis>
  <nanos>114500</nanos>
  <sequence>530</sequence>
  <logger>eubos.main</logger>
  <level>INFO</level>
  <class>eubos.main.EubosEngineMain</class>
  <method>sendBestMoveCommand</method>
  <thread>39</thread>
  <message>Search result: pv=f8e8:r:best e7d6:B:p:best a5b6:q e1g1:K:best e8d8:r d2f4:Q:best d8d6:r:B:best f4g4:Q:n:best b8c6:n:best g4h3:Q c6d4:n:best e2f4:N:best d4b3:n:B , score=0 mate=false, depth=13, trusted=0 rootTrans=(trans best=f8e8:r:best, dep=13, sc=-8, type=2 age=3)</message>
</record>
<record>
  <date>2023-12-07T21:14:46.919114500Z</date>
  <millis>1701983686919</millis>
  <nanos>114500</nanos>
  <sequence>531</sequence>
  <logger>eubos.main</logger>
  <level>INFO</level>
  <class>eubos.main.EubosEngineMain</class>
  <method>convertToGenericAndSendBestMove</method>
  <thread>39</thread>
  <message>Best Move g7h8</message>
</record>
``` 

This is indicating that the hash returned from the search and the principal variation identify Re8 as the best move, but somehow the function that wrote it out to the GUI sent Bh8??

There is only one way this could have happened, from analysing the code, and that is that a bad hash entry overwrote the good one, on shutdown of the search. I hope to pin this down and fix it in Eubos v2.27, which is forthcoming.

**Comment by cjbolt** [comment]
And reproduced again:
`5rk1/p2b2rp/1p2pqp1/3p4/2pP4/P4BQN/2P1RPPP/4R1K1 b - - - 31`

selected Nd4 despite identifying Ne7 as the best move!!!
```
<record>
  <date>2023-12-07T22:01:07.643566600Z</date>
  <millis>1701986467643</millis>
  <nanos>566600</nanos>
  <sequence>12493</sequence>
  <logger>eubos.main</logger>
  <level>INFO</level>
  <class>eubos.main.EubosEngineMain</class>
  <method>sendBestMoveCommand</method>
  <thread>103</thread>
  <message>Search result: pv=c6e7:n:best , score=0 mate=false, depth=16, trusted=0 rootTrans=(trans best=c6e7:n:best, dep=15, sc=46, type=3 age=7)</message>
</record>
<record>
  <date>2023-12-07T22:01:07.823575500Z</date>
  <millis>1701986467823</millis>
  <nanos>575500</nanos>
  <sequence>12496</sequence>
  <logger>eubos.main</logger>
  <level>INFO</level>
  <class>eubos.main.EubosEngineMain</class>
  <method>convertToGenericAndSendBestMove</method>
  <thread>103</thread>
  <message>Best Move c6d4</message>
</record>
```

**Comment by cjbolt** [comment]
Fixed in Eubos v2.27
